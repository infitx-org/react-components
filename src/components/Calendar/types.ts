export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
