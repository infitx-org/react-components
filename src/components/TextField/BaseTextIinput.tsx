import React from "react";
import "./TextField.scss";

export interface BaseTextFieldProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  value?: string;
}

function BaseTextField(
  { ...props }: BaseTextFieldProps,
  ref: React.Ref<HTMLInputElement>
): JSX.Element {
  return <input {...props} type="text" ref={ref} />;
}

export default React.forwardRef(React.memo(BaseTextField));
