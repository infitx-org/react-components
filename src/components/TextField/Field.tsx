import React from "react";
import classnames from "classnames";
import { Kind } from "../types";

export interface FieldPublicProps {
  kind: `${Kind}`;
  label?: string;
}

interface FieldPrivateProps {
  focused: boolean;
  children: React.ReactNode;
}

type FieldProps = FieldPublicProps & FieldPrivateProps;

export default function Field({
  kind,
  label,
  focused,
  children,
}: FieldProps): JSX.Element {
  const fieldClassname = classnames([
    "field",
    kind && `field--${kind}`,
    focused && "field--focused",
  ]);
  return (
    <div className={fieldClassname}>
      <label>{label}</label>
      {children}
    </div>
  );
}
