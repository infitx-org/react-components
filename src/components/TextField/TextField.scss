@use "../../scss/base/colors.scss";
@use "../../scss/mixins/field.scss";
@use "../../scss/mixins/any.scss";

.rc-textfield {
  @include any.rc-component;

  flex: 2 1 auto;
  vertical-align: middle;
  position: relative;
  height: 20px;
  background: transparent;
  border: none;
  outline: none;
  padding: 0;
  pointer-events: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: colors.$dark-gray;
}
