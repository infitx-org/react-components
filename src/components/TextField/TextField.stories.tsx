import TextField from "./TextField";

export default {
  title: "Components/TextField",
};

/* eslint-disable no-console */

const Template = (args) => <TextField {...args} />;

export const Default = Template.bind({});
Default.args = {
  kind: "primary",
  size: "large",
  value: "test",
  disabled: false,
  className: undefined,
  placeholder: "Type something",
  required: false,
  invalid: false,
  pending: false,
  // eslint-disable-next-line
  onChange: console.log,
  validation: [
    { active: false, message: "Test" },
    { active: true, message: "invalid" },
    { active: undefined, message: "unknown" },
  ],
};

export const TypePassword = Template.bind({});
TypePassword.args = {
  type: "password",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: "Placeholder...",
};

export const Pending = Template.bind({});
Pending.args = {
  pending: true,
};

export const Required = Template.bind({});
Required.args = {
  required: true,
};

export const Invalid = Template.bind({});
Invalid.args = {
  invalid: true,
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};
