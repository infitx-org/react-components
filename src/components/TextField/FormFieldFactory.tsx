import React from "react";
import Field, { FieldPublicProps } from "./Field";
import Spinner from "../Spinner";

type InputRef = React.ForwardedRef<HTMLInputElement>;
type FuncRef = (node: HTMLInputElement) => void;

const mergeRefs = (...refs: (InputRef | FuncRef)[]) => {
  const filteredRefs = refs.filter(Boolean);
  if (!filteredRefs.length) return null;
  if (filteredRefs.length === 0) return filteredRefs[0];
  return (element: HTMLInputElement) => {
    filteredRefs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(element);
      } else if (ref) {
        // eslint-disable-next-line
        ref.current = element;
      }
    });
  };
};

export type FormFieldFactoryProps<T> = T & FieldPublicProps;

export default function FormFieldFactory<T>(
  Component: React.FunctionComponent<T>
) {
  return React.forwardRef<HTMLInputElement, FormFieldFactoryProps<T>>(
    function ForwardWrapperRefs(props: FormFieldFactoryProps<T>, forwardedRef) {
      const inputRef = React.useRef<HTMLInputElement>(null);
      const [focused, setFocused] = React.useState(false);
      const setFocus = () => setFocused(true);
      const unsetFocus = () => setFocused(false);

      React.useEffect(() => {
        inputRef.current?.addEventListener("focus", setFocus);
        inputRef.current?.addEventListener("blur", unsetFocus);
        return () => {
          inputRef.current?.removeEventListener("focus", setFocus);
          inputRef.current?.removeEventListener("blur", unsetFocus);
        };
      }, [inputRef.current]);

      const { kind, label, ...componentProps } = props;

      return (
        <Field kind={kind} focused={focused} label={label}>
          <Component
            {...((componentProps as unknown) as T)}
            inputRef={mergeRefs(inputRef, forwardedRef)}
          />
        </Field>
      );
    }
  );
}

export type FormFieldStatusProps<T> = T & { pending: boolean };

export function FormFieldStatus<T>(Component: React.FunctionComponent<T>) {
  return function WithStatus({ pending, ...props }: FormFieldStatusProps<T>) {
    return <Component {...props}>{pending && <Spinner size="s" />}</Component>;
  };
}
