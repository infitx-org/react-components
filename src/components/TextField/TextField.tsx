import React from "react";
import classnames from "classnames";
import { Kind, InputSize, KeyCode } from "types";
import Field, { Loader, Placeholder, InvalidIcon } from "components/Field";
import "./TextField.scss";

type BaseInput = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "size" | "onChange"
>;

export interface TextFieldProps extends BaseInput {
  kind?: `${Kind}`;
  size?: `${InputSize}`;
  type?: "text" | "password";
  className?: string;
  placeholder?: string;
  value?: string;
  required?: boolean;
  invalid?: boolean;
  pending?: boolean;
  onChange?: (value: string) => void;
}

export default React.forwardRef(function TextField(
  {
    kind = Kind.Primary,
    size = InputSize.Large,
    type = "text",
    className,
    placeholder,
    value = "",
    required,
    invalid,
    pending,
    onChange,
    ...props
  }: TextFieldProps,
  forwardedRef: React.ForwardedRef<HTMLDivElement>
): JSX.Element {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [currentValue, setValue] = React.useState<string | undefined>(value);
  const [open, setOpen] = React.useState(false);
  const [focused, setFocused] = React.useState(false);

  React.useEffect(() => {
    setValue(value || "");
  }, [value]);

  function enter() {
    setFocused(true);
    setOpen(true);
    inputRef.current?.focus();
  }

  function leave() {
    setFocused(false);
    setOpen(false);
  }

  async function onValueChange(e: React.ChangeEvent<HTMLInputElement>) {
    setValue(e.target.value);
    onChange?.(e.target.value);
  }

  function onFocus(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onFocus?.(e);
      enter();
    }
  }

  function onBlur(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onBlur?.(e);
    }
  }

  function onFieldClick(): void {
    if (!open) {
      enter();
    }
    inputRef.current?.focus();
  }

  function onKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    const { keyCode } = e;

    if (keyCode === KeyCode.Tab) {
      leave();
    }
  }

  const textFieldClassName = classnames([
    "rc-textfield",
    `rc-textfield--${size}`,
  ]);

  return (
    <Field
      className={className}
      kind={kind}
      size={size}
      required={required && currentValue === ""}
      invalid={invalid}
      disabled={props.disabled}
      focused={focused}
      onClick={onFieldClick}
      onClickOutside={leave}
      ref={forwardedRef}
    >
      {placeholder && (
        <Placeholder
          label={placeholder}
          active={currentValue !== "" || focused}
          size={size}
        />
      )}
      <input
        {...props}
        value={currentValue}
        className={textFieldClassName}
        type={type}
        ref={inputRef}
        onChange={onValueChange}
        onFocus={onFocus}
        onBlur={onBlur}
        onKeyDown={onKeyDown}
      />
      {pending && <Loader size={size} />}
      {invalid && <InvalidIcon size={size} />}
    </Field>
  );
});
