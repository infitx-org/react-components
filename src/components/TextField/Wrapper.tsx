import React from "react";
import Field, { FieldPublicProps } from "./Field";

type InputRef = React.ForwardedRef<HTMLInputElement>;
type FuncRef = (node: HTMLInputElement) => void;

const mergeRefs = (...refs: (InputRef | FuncRef)[]) => {
  const filteredRefs = refs.filter(Boolean);
  if (!filteredRefs.length) return null;
  if (filteredRefs.length === 0) return filteredRefs[0];
  return (element: HTMLInputElement) => {
    filteredRefs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(element);
      } else if (ref) {
        // eslint-disable-next-line
        ref.current = element;
      }
    });
  };
};

export default function Wrapper<T>(Component: React.FunctionComponent<T>) {
  // const ComponentWithForwardedRefs = React.forwardRef<HTMLInputElement, T>(
  //   (props, ref) => <Component {...props} inputRef={ref} />
  // );

  return React.forwardRef<HTMLInputElement, WrapperProps<T>>(
    function ForwardWrapperRefs(props: WrapperProps<T>, forwardedRef) {
      const inputRef = React.useRef<HTMLInputElement>(null);
      const [focused, setFocused] = React.useState(false);
      const setFocus = () => setFocused(true);
      const unsetFocus = () => setFocused(false);

      React.useEffect(() => {
        inputRef.current?.addEventListener("focus", setFocus);
        inputRef.current?.addEventListener("blur", unsetFocus);
        return () => {
          inputRef.current?.removeEventListener("focus", setFocus);
          inputRef.current?.removeEventListener("blur", unsetFocus);
        };
      }, [inputRef.current]);

      return (
        <Field kind={props.kind} focused={focused} label={props.label}>
          <Component
            {...(props as T)}
            inputRef={mergeRefs(inputRef, forwardedRef)}
          />
        </Field>
      );
    }
  );
}

export type WrapperProps<T> = T & FieldPublicProps;
