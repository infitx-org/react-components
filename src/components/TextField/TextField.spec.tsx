import { render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { Kind, InputSize } from "types";
import TextField from "./TextField";

const testFile = new File(
  [new Blob(["test"], { type: "text/plain" })],
  "test.dat"
);
const commonProps = {
  onChange: jest.fn(),
};

function getInput(container: HTMLElement): HTMLInputElement {
  return container.querySelector("input") as HTMLInputElement;
}

describe("tests the TextField props", () => {
  it("renders the TextField", () => {
    const { container } = render(<TextField {...commonProps} />);
    expect(container.querySelector(".rc-textfield")).toBeTruthy();
    expect(container.querySelectorAll("input")).toHaveLength(1);
  });

  it("renders the value", () => {
    const { container } = render(<TextField {...commonProps} value="test" />);
    const input = getInput(container);
    expect(input.value).toBe("test");
  });

  it("renders empty string when value is not specified", () => {
    const { container } = render(<TextField {...commonProps} />);
    const input = getInput(container);
    expect(input.value).toBe("");
  });

  it("renders the placeholder", () => {
    const { container } = render(
      <TextField {...commonProps} placeholder="test" />
    );
    expect(container.querySelector(".placeholder")).toBeTruthy();
  });

  it("renders the classname", () => {
    const { container } = render(
      <TextField {...commonProps} className="test" />
    );
    expect(container.querySelector(".test")).toBeTruthy();
  });

  it("renders as disabled", () => {
    const { container } = render(<TextField {...commonProps} disabled />);
    const input = getInput(container);
    expect(input.disabled).toBeTruthy();
  });

  it("renders as pending", () => {
    const { container } = render(<TextField {...commonProps} pending />);
    expect(container.querySelector(".rc-spinner")).toBeTruthy();
  });

  it("renders as required", () => {
    const { container } = render(<TextField {...commonProps} required />);
    expect(container.querySelector(".rc-field--required")).toBeTruthy();
  });

  it("renders as invalid", () => {
    const { container } = render(<TextField {...commonProps} invalid />);
    expect(container.querySelector(".rc-field--invalid")).toBeTruthy();
    expect(container.querySelector(".rc-invalid-icon")).toBeTruthy();
  });

  it("renders the small, medium, large sizes", () => {
    Object.values(InputSize).forEach((size) => {
      const { container } = render(<TextField {...commonProps} size={size} />);
      expect(container.querySelector(`.rc-field--${size}`)).toBeTruthy();
    });
  });

  it("renders all the kinds", () => {
    Object.values(Kind).forEach((kind) => {
      const { container } = render(<TextField {...commonProps} kind={kind} />);
      expect(container.querySelector(`.rc-field--${kind}`)).toBeTruthy();
    });
  });

  it("triggers onFocus when focusing", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <TextField {...commonProps} onFocus={mockFn} />
    );
    fireEvent.focus(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when clicking", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <TextField {...commonProps} onFocus={mockFn} />
    );
    userEvent.click(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when tabbing", () => {
    const mockFn = jest.fn();
    render(<TextField {...commonProps} onFocus={mockFn} />);
    userEvent.tab();
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onBlur when clicking out", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <TextField {...commonProps} onBlur={mockFn} />
    );
    userEvent.click(getInput(container));
    userEvent.click(document.body);
    expect(mockFn).toHaveBeenCalled();
  });

  // it("triggers onBlur when tabbing away", () => {
  //   const mockFn = jest.fn();
  //   const { container } = render(<TextField {...commonProps} onBlur={mockFn} />);
  //   userEvent.click(getInput(container));
  //   userEvent.tab();
  //   expect(mockFn).toHaveBeenCalled();
  // });

  it("trigger onChange", async () => {
    const mockFn = jest.fn();
    const { container } = render(<TextField onChange={mockFn} />);
    fireEvent.change(getInput(container), { target: { value: "test" } });
    expect(getInput(container).value).toBe("test");
    expect(mockFn).toHaveBeenCalledWith("test");
  });
});

// Snapshot testing
it("renders the TextField correctly when multiple props are set", () => {
  const { container } = render(
    <TextField {...commonProps} placeholder="test" />
  );
  expect(container).toMatchSnapshot();
});
