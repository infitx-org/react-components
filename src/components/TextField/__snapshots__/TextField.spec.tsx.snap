// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the TextField correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-field rc-field--primary rc-field--large rc-field--large"
    role="presentation"
  >
    <label
      class="placeholder placeholder--large"
    >
      test
    </label>
    <input
      class="rc-textfield rc-textfield--large"
      type="text"
      value=""
    />
  </div>
</div>
`;
